File: 138.CopyListRandomPointer/copyRandomList.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        
        # Step 1: Create a copy of each node and insert it next to the original node
        current = head
        while current:
            new_node = Node(current.val)
            new_node.next = current.next
            current.next = new_node
            current = new_node.next
        
        # Step 2: Set the random pointers for the copied nodes
        current = head
        while current:
            if current.random:
                current.next.random = current.random.next
            current = current.next.next
        
        # Step 3: Separate the original and copied nodes
        original = head
        copy_head = head.next
        copy_current = copy_head
        
        while original:
            original.next = original.next.next if original.next else None
            copy_current.next = copy_current.next.next if copy_current.next else None
            
            original = original.next
            copy_current = copy_current.next
        
        return copy_head

File: 138.CopyListRandomPointer/copyRandomList.ts
/**
 * Definition for _Node.
 * class _Node {
 *     val: number
 *     next: _Node | null
 *     random: _Node | null
 *
 *     constructor(val?: number, next?: _Node, random?: _Node) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *         this.random = (random===undefined ? null : random)
 *     }
 * }
 */

function copyRandomList(head: _Node | null): _Node | null {
    if (!head) return null;

    const map = new Map<_Node, _Node>();

    // First pass: create a copy of each node and store it in the map
    let current = head;
    while (current) {
        map.set(current, new _Node(current.val));
        current = current.next;
    }

    // Second pass: assign next and random pointers
    current = head;
    while (current) {
        const copiedNode = map.get(current);
        copiedNode!.next = map.get(current.next) || null;
        copiedNode!.random = map.get(current.random) || null;
        current = current.next;
    }

    return map.get(head) || null;
}